EpikCord/client/websocket_client.py

```diff
@@ -48,198 +58,231 @@ def __init__(
-        elif self.ws.close_code == GatewayCECode.ShardingRequired:
+        elif self.websocket.close_code == GatewayCECode.ShardingRequired:
             raise ShardingRequired("You need to shard the bot.")
-        elif self.ws.close_code == GatewayCECode.InvalidAPIVersion:
+        elif self.websocket.close_code == GatewayCECode.InvalidAPIVersion:
             raise DeprecationWarning(
                 "The gateway you're connecting to is deprecated and does not "
                 "work, upgrade EpikCord.py. "
             )
-        elif self.ws.close_code == GatewayCECode.InvalidIntents:
+        elif self.websocket.close_code == GatewayCECode.InvalidIntents:
             raise InvalidIntents("The intents you provided are invalid.")
-        elif self.ws.close_code == GatewayCECode.UnknownError:
+        elif self.websocket.close_code == GatewayCECode.UnknownError:
             await self.resume()
-        elif self.ws.close_code == GatewayCECode.UnknownOpcode:
+        elif self.websocket.close_code == GatewayCECode.UnknownOpcode:
             logger.critical(
                 "EpikCord.py sent an invalid OPCODE to the Gateway. "
                 "Report this immediately. "
             )
             await self.resume()
-        elif self.ws.close_code == GatewayCECode.DecodeError:
+        elif self.websocket.close_code == GatewayCECode.DecodeError:
             logger.critical(
                 "EpikCord.py sent an invalid payload to the Gateway."
                 " Report this immediately. "
             )
             await self.resume()
-        elif self.ws.close_code == GatewayCECode.NotAuthenticated:
+        elif self.websocket.close_code == GatewayCECode.NotAuthenticated:
             logger.critical(
                 "EpikCord.py has sent a payload prior to identifying."
                 " Report this immediately. "
             )
 
-        elif self.ws.close_code == GatewayCECode.AlreadyAuthenticated:
+        elif self.websocket.close_code == GatewayCECode.AlreadyAuthenticated:
             logger.critical(
                 "EpikCord.py tried to authenticate again." " Report this immediately. "
             )
             await self.resume()
-        elif self.ws.close_code == GatewayCECode.InvalidSequence:
+        elif self.websocket.close_code == GatewayCECode.InvalidSequence:
             logger.critical(
                 "EpikCord.py sent an invalid sequence number."
                 " Report this immediately."
             )
             await self.resume()
-        elif self.ws.close_code == GatewayCECode.SessionTimeout:
+        elif self.websocket.close_code == GatewayCECode.SessionTimeout:
             logger.critical("Session timed out.")
             await self.resume()
         else:
             raise ClosedWebSocketConnection(
-                f"Connection has been closed with code {self.ws.close_code}"
+                f"Connection has been closed with code {self.websocket.close_code}"
             )
```

Comment:
I believe that the `handle_close` conditionals branches either raise an exception or resume with optional logging message.

There is something to be done here to reduce the complexity of this method, and provide a dictionary of behavior that can be handle in a generic manner.

Reply:
They log specifically what happened, but yes it does look rather complex, and maybe just a dict of functions will do, in the mapping [int, Callback]

Reply:
Well as said there is only 2 distinct behavior, and it wouldn't be efficient to have a callback for each. 

I was more thinking something such as a state object rather than a callback.
That way:

```py
exit_behavior = self._close_behavior[self.websocket.close_code]

if not exit_behavior:
    raise ...  # Not Implemented, thus a new kind of close that has  to be issued

if exit_behavior.type == UNRECOVERABLE:
    raise ...

elif exit_behavior == CRITICAL:
   log.critical(...)

if exit_behavior.can_be_resumed:
    await self.resume()
```